File: ModelGovernance/FLASKMODELGOVERNANCE/flaskr/additional_scripts/PREDICT.py
"""Makes prediction using given model and data
"""

import warnings

warnings.filterwarnings("ignore", category=FutureWarning)

import sys
import pickle
import pandas as pd
import numpy as np

model = sys.argv[1]  # name fo the model
date = sys.argv[2]  # timestamp
type = sys.argv[3]  # type of prediction
is_hash = sys.argv[4]  # False if hash was not provided

# load model
with open("flaskr/V/Models/" + model + "/model", 'rb') as fd:
    model = pickle.load(fd)

# loading data
if is_hash == '0':
    # case when hash was not provided
    # reading data
    data = pd.read_csv("flaskr/tmp/" + date + ".csv", delimiter=',', header=0)
else:
    # case when hash was not provided
    # reading hash
    hash = sys.argv[5]
    # reading target column name
    target = sys.argv[6]

    # reading data
    data = pd.read_csv("flaskr/V/Datasets/" + hash, delimiter=',', header=0)
    # dropping target column
    data = data.drop(columns=target)

# making prediction
if type == 'prob':
    pred = model.predict_proba(data)
elif type == 'exact':
    pred = model.predict(data)

# saving result
with open("flaskr/tmp/" + date + ".txt", "wb") as fd:
    np.savetxt(fd, pred, delimiter=',')

File: docker/ModelGovernance/FLASKMODELGOVERNANCE/flaskr/additional_scripts/PREDICT.r
# command line arguments
args = commandArgs(trailingOnly=TRUE)

# model name
model = args[1]
# timestamp
date = args[2]
# type of the prediction
type = args[3]
# flag if hash was provided
is_hash = args[4]

# reading model
model = readRDS(paste0('../../../V/Models/', model, '/model'))

if(is_hash == '1') {
	# case when hash was provided
	# hash
	hash = args[5]
	# target column
	target = args[6]
	# reading data
	data = read.csv(paste0('../../../V/Datasets/', hash), header=T)
	# droping target
	data = data[-which(colnames(data) == target)]
} else {
	# case when hash was not provided
	# reading the data
	data = read.csv(paste0('../../../tmp/', date, '.csv'), header=T)
}

# making predictions
if(class(model)[1] == 'WrappedModel') {
	# case when model is mlr model
	library(mlr)
	result = predict(model, newdata = data)$data$response
} else if(class(model)[1] == 'train' && class(model)[2] == 'train.formula') {
	# case when model is caret model
	result = predict(model, data)
} else if(class(model)[2] == 'model_fit') {
	# case when model is parsnip model
	result = parsnip::predict(model, data)
}

# writting result
write.table(result, paste0('../../../tmp/', date, '.txt'), sep=',', col.names=F, row.names = F)

